File: frontend/src/services/auditEntityLicense.ts
export type AuditEntityLicenseType = {
  id: number;
  license_type_code: string;
  license_type_name: string;
  description: string | null;
  is_active: boolean;
  created_by: string | null;
  updated_by: string | null;
  created_at: string;
  updated_at: string;
};

export type AuditEntityLicenseTypeListResponse = {
  total: number;
  items: AuditEntityLicenseType[];
};

export type AuditEntityLicenseStatus =
  | "valid"
  | "expired"
  | "pending"
  | "suspended"
  | "cancelled"
  | "not_applicable";

export type AuditEntityLicense = {
  id: number;
  audit_entity_id: number;
  license_type_id: number;
  license_no: string;
  license_name: string | null;
  issuing_authority: string | null;
  issuing_country: string | null;
  issue_date: string | null;
  expiry_date: string | null;
  renewal_due_date: string | null;
  license_status: AuditEntityLicenseStatus;
  document_reference: string | null;
  is_mandatory: boolean;
  is_verified: boolean;
  remarks: string | null;
  is_active: boolean;
  created_by: string | null;
  updated_by: string | null;
  created_at: string;
  updated_at: string;
};

export type AuditEntityLicenseListResponse = {
  total: number;
  page: number;
  page_size: number;
  items: AuditEntityLicense[];
};

export type AuditEntityLicensePayload = {
  audit_entity_id: number;
  license_type_id: number;
  license_no: string;
  license_name?: string | null;
  issuing_authority?: string | null;
  issuing_country?: string | null;
  issue_date?: string | null;
  expiry_date?: string | null;
  renewal_due_date?: string | null;
  license_status: AuditEntityLicenseStatus;
  document_reference?: string | null;
  is_mandatory: boolean;
  is_verified: boolean;
  remarks?: string | null;
};

export type AuditEntityLicenseUpdatePayload =
  Partial<AuditEntityLicensePayload> & {
    is_active?: boolean;
  };

type ListParams = {
  page: number;
  pageSize: number;
  search?: string;
  isActive?: boolean;
  auditEntityId?: number;
  licenseTypeId?: number;
  licenseStatus?: AuditEntityLicenseStatus;
  isMandatory?: boolean;
  isVerified?: boolean;
  sortBy?: string;
  sortOrder?: "asc" | "desc";
};

function buildQuery(params: ListParams) {
  const query = new URLSearchParams();

  query.set("page", String(params.page));
  query.set("page_size", String(params.pageSize));
  query.set("sort_by", params.sortBy ?? "id");
  query.set("sort_order", params.sortOrder ?? "asc");

  if (params.search) query.set("search", params.search);

  if (typeof params.isActive === "boolean") {
    query.set("is_active", String(params.isActive));
  }

  if (params.auditEntityId) {
    query.set("audit_entity_id", String(params.auditEntityId));
  }

  if (params.licenseTypeId) {
    query.set("license_type_id", String(params.licenseTypeId));
  }

  if (params.licenseStatus) {
    query.set("license_status", params.licenseStatus);
  }

  if (typeof params.isMandatory === "boolean") {
    query.set("is_mandatory", String(params.isMandatory));
  }

  if (typeof params.isVerified === "boolean") {
    query.set("is_verified", String(params.isVerified));
  }

  return query.toString();
}

async function requestJson<T>(
  url: string,
  options?: RequestInit,
): Promise<T> {
  const response = await fetch(url, {
    ...options,
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      ...(options?.headers ?? {}),
    },
  });

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    const message =
      error?.detail || error?.message || "Audit entity license request failed.";

    throw new Error(message);
  }

  return (await response.json()) as T;
}

export async function listAuditEntityLicenseTypes(isActive = true) {
  return requestJson<AuditEntityLicenseTypeListResponse>(
    `/api/backend/audit-entity-licenses/types?is_active=${String(isActive)}`,
    {
      method: "GET",
    },
  );
}

export async function listAuditEntityLicenses(params: ListParams) {
  const query = buildQuery(params);

  return requestJson<AuditEntityLicenseListResponse>(
    `/api/backend/audit-entity-licenses?${query}`,
    {
      method: "GET",
    },
  );
}

export async function createAuditEntityLicense(
  payload: AuditEntityLicensePayload,
) {
  return requestJson<{
    message: string;
    data: AuditEntityLicense;
  }>("/api/backend/audit-entity-licenses", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function updateAuditEntityLicense(
  id: number,
  payload: AuditEntityLicenseUpdatePayload,
) {
  return requestJson<{
    message: string;
    data: AuditEntityLicense;
  }>(`/api/backend/audit-entity-licenses/${id}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export async function deactivateAuditEntityLicense(id: number) {
  return requestJson<{
    message: string;
    data: AuditEntityLicense;
  }>(`/api/backend/audit-entity-licenses/${id}`, {
    method: "DELETE",
  });
}

export async function restoreAuditEntityLicense(id: number) {
  return requestJson<{
    message: string;
    data: AuditEntityLicense;
  }>(`/api/backend/audit-entity-licenses/${id}/restore`, {
    method: "PATCH",
  });
}

export async function permanentDeleteAuditEntityLicense(id: number) {
  return requestJson<{ message: string; data: null }>(
    `/api/backend/audit-entity-licenses/${id}/permanent`,
    {
      method: "DELETE",
    },
  );
}

File: backend/app/schemas/audit_entity_license.py
from datetime import date, datetime

from pydantic import BaseModel, Field


class AuditEntityLicenseTypeBase(BaseModel):
    license_type_code: str = Field(..., max_length=20)
    license_type_name: str = Field(..., max_length=150)
    description: str | None = None


class AuditEntityLicenseTypeResponse(AuditEntityLicenseTypeBase):
    id: int
    is_active: bool
    created_by: str | None = None
    updated_by: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True,
    }


class AuditEntityLicenseTypeListResponse(BaseModel):
    total: int
    items: list[AuditEntityLicenseTypeResponse]


class AuditEntityLicenseBase(BaseModel):
    audit_entity_id: int
    license_type_id: int
    license_no: str = Field(..., max_length=150)
    license_name: str | None = Field(default=None, max_length=200)
    issuing_authority: str | None = Field(default=None, max_length=200)
    issuing_country: str | None = Field(default="Bangladesh", max_length=100)
    issue_date: date | None = None
    expiry_date: date | None = None
    renewal_due_date: date | None = None
    license_status: str = Field(default="valid", max_length=30)
    document_reference: str | None = Field(default=None, max_length=255)
    is_mandatory: bool = False
    is_verified: bool = False
    remarks: str | None = None


class AuditEntityLicenseCreate(AuditEntityLicenseBase):
    pass


class AuditEntityLicenseUpdate(BaseModel):
    audit_entity_id: int | None = None
    license_type_id: int | None = None
    license_no: str | None = Field(default=None, max_length=150)
    license_name: str | None = Field(default=None, max_length=200)
    issuing_authority: str | None = Field(default=None, max_length=200)
    issuing_country: str | None = Field(default=None, max_length=100)
    issue_date: date | None = None
    expiry_date: date | None = None
    renewal_due_date: date | None = None
    license_status: str | None = Field(default=None, max_length=30)
    document_reference: str | None = Field(default=None, max_length=255)
    is_mandatory: bool | None = None
    is_verified: bool | None = None
    remarks: str | None = None
    is_active: bool | None = None


class AuditEntityLicenseResponse(AuditEntityLicenseBase):
    id: int
    is_active: bool
    created_by: str | None = None
    updated_by: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True,
    }


class AuditEntityLicenseListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[AuditEntityLicenseResponse]


class AuditEntityLicenseMessageResponse(BaseModel):
    message: str
    data: AuditEntityLicenseResponse | None = None
